### PR TITLE
optimze: reuse subset entry to optimze alloc/inuse memory

### PR DESCRIPTION
**Pull request by dzdx**
### Issues associated with this PR

issue: https://github.com/mosn/mosn/issues/2111
Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

#### before

```
  ✘ dzdx@B-QBDRMD6M-0201  ~/Documents/mosn-open/pkg/upstream/cluster   subset-lb-iface ±  go test -bench BenchmarkSubsetLoadBalancer/subsetLoadBalancerPreIndex -benchmem  -run=^$ 
2022-08-29 14:57:30,180 [INFO] [network] [ register pool factory] register protocol: mock factory
goos: darwin
goarch: amd64
pkg: mosn.io/mosn/pkg/upstream/cluster
BenchmarkSubsetLoadBalancer/subsetLoadBalancerPreIndex-12                    256           5013841 ns/op         3344313 B/op       1956 allocs/op
PASS
ok      mosn.io/mosn/pkg/upstream/cluster       2.061s
```
#### after

```
 dzdx@B-QBDRMD6M-0201  ~/Documents/mosn-open/pkg/upstream/cluster   subset-lb-iface  go test -bench BenchmarkSubsetLoadBalancer/subsetLoadBalancerPreIndex -benchmem  -run=^$ 
2022-08-29 14:57:56,611 [INFO] [network] [ register pool factory] register protocol: mock factory
goos: darwin
goarch: amd64
pkg: mosn.io/mosn/pkg/upstream/cluster
BenchmarkSubsetLoadBalancer/subsetLoadBalancerPreIndex-12                    514           2605042 ns/op          653241 B/op       1910 allocs/op
PASS
ok      mosn.io/mosn/pkg/upstream/cluster       2.026s
```

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/upstream/cluster/subset_loadbalancer_builder.go [hunk 1]
```diff
@@ -121,15 +123,19 @@ func (b *subsetLoadBalancerBuilder) createSubsets() types.LbSubsetMap {
 }
 
 func (b *subsetLoadBalancerBuilder) selectHosts(s *intsets.Sparse) []types.Host {
-	if s == nil {
-		return nil
+	if s == nil || s.IsEmpty() {
+		return make([]types.Host, 0)
```
Comment (nejisama) [inline pkg/upstream/cluster/subset_loadbalancer_builder.go:127]: 这里从nil 改成非nil 了 会不会有问题

File: pkg/upstream/cluster/subset_loadbalancer_builder.go [hunk 2]
```diff
@@ -272,3 +278,60 @@ func subsetMergeKeys(subSetKeys []types.SortedStringSetType, defaultSubset types
...
+type sparseEntry struct {
+	key   *intsets.Sparse
+	value interface{}
+}
+
+func (e sparseEntry) HashCode() int {
+	min := e.key.Min()
+	max := e.key.Max()
+	length := e.key.Len()
+	return min<<44 | max<<22 | length
```
Comment (nejisama) [inline pkg/upstream/cluster/subset_loadbalancer_builder.go:323]: 这个HASH的原理是什么

File: pkg/upstream/cluster/subset_loadbalancer_builder.go (hunk 1 above)
Comment (chuailiwu) [reply pkg/upstream/cluster/subset_loadbalancer_builder.go:127]: 看了下实现 非nil不会有问题，直接返回nil可能会有问题，很多逻辑没有判空

File: pkg/upstream/cluster/subset_loadbalancer_builder.go (hunk 2 above)
Comment (chuailiwu) [reply pkg/upstream/cluster/subset_loadbalancer_builder.go:323]: 看了下就是min、max和len分别占用三段位子组成一个int，如果hash碰撞了会有数组来解决

**Review by chuailiwu — APPROVED** [review]
lgtm

File: pkg/upstream/cluster/subset_loadbalancer_builder.go (hunk 2 above)
Comment (dzdx) [reply pkg/upstream/cluster/subset_loadbalancer_builder.go:323]: 是的，就是类似hashmap的解法，min、max、len不一样的话hosts肯定不一样

**Review by nejisama — APPROVED** [review]

**Review by 3062 — APPROVED** [review]

**Review by taoyuanyuan — APPROVED** [review]
